proposals/NNNN-DXIL-Data-Scalarization.md

```diff
@@ -0,0 +1,62 @@
+# DXIL Data Scalarization
+
+* Proposal: [NNNN](NNNN-DXIL-Data-Scalarization.md)
+* Author(s): [Farzon Lotfi](https://github.com/farzonl)
+* Sponsor: [Farzon Lotfi](https://github.com/farzonl)
+* Status: **Under Consideration**
+
+## Introduction
+
+In Proposal [0009](0009-DXIL-Function-Scalarization.md) we covered scalarization
+of call instructions, and vector operations like math ops, logical ops,
+bitcasts, loads, and stores. The goal of this proposal is to present a solution
+for scalarizing data structures via data layout transformations. As a note,
+any case that results in an implicit cbuffer will not be covered by this
+proposal.
+
+## Motivation
+
+As mentioned in [DXIL.rst](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#vectors)
+"HLSL vectors are scalarized" and "Matrices are lowered to vectors". Therefore,
+we need to be able to scalarize these types in clang DirectX backend. Without
+scalarizing the data structures and call instructions we can't generate legal
+DXIL.
+
+## Background
+
+In DXC we support five forms of data scalarization. data marked with
+`groupshared`, `static` function scope, `static` global scope, global data,
+and data in function scope. There are then subset behaviors depending on
+if we are dealing with arrays of vectors or vectors. The data
+scalarization can then be broken down into three cases:
+
+1. `static` scalar layouts
+   * vector of `M` size get scalarized into `M` elements.
+   * an array of `N` size with a vector of `M` will get scalarized into `M`
```

Comment:
Is this referring to an array of N vectors, each of size M?
If so, then why wouldn't it instead be scalarized into N arrays of M size? Or is that semantically equivalent?
I would think the dimension of the higher-dimensional structure would stay in its place as the outer dimension.

Reply:
Thats just how DXC does it. We aren't going to carry forward that implementation detail.